pipelines/zendesk/__init__.py

```diff
@@ -0,0 +1,221 @@
+    """
+    The source for the dlt pipeline. It returns all the basic information.
+    @:param credentials: read as a dict, as filled in .dlt.secrets.toml
+    @:returns: multiple dlt resources
+    """
+    # Authenticate
+    zendesk_client = auth_zenpy(credentials=credentials)
+    return chats_table(zendesk_client=zendesk_client)
+
+
+@dlt.source
+def zendesk(credentials: Union[ZendeskCredentialsOAuth, ZendeskCredentialsToken, ZendeskCredentialsEmailPass] = dlt.secrets.value, load_all: bool = True, pivot_ticket_fields: bool = True,
+            incremental_start_time: DateTime = FIRST_DAY_OF_MILLENNIUM) -> Sequence[DltResource]:
+    """
+    The source for the dlt pipeline. It returns all the basic tables for Zendesk Support: tickets, users, brands, organizations, groups and all extra resources if required
+    @:param credentials: read as a dict, as filled in .dlt.secrets.toml
+    @:param load_all: bool, if true will load extra resources for the api
+    @:param pivot_tickets: bool that indicates whether to pivot the custom fields in tickets, default set to true
+    @:param incremental_start_time: Datetime taht indicates when to start loading objects that use incremental loading
+    @:returns: multiple dlt resources
+    """
+
+    # TODO: Make caching manageable and editable by users
+    # Authenticate
+    zendesk_client = auth_zenpy(credentials=credentials)
+    resource_list = [ticket_fields_table(zendesk_client=zendesk_client),
+                     ticket_table(zendesk_client=zendesk_client, pivot_fields=pivot_ticket_fields, start_time=incremental_start_time),
+                     ticket_metric_table(zendesk_client=zendesk_client, start_time=incremental_start_time)]
+    # load other api endpoints into resources if the option is to do so
+    if load_all:
+        resources_to_be_loaded.extend(EXTRA_RESOURCES)
+    for resource in resources_to_be_loaded:
+        resource_list.append(dlt.resource(basic_resource(zendesk_client=zendesk_client, resource=resource),
+                                          name=resource,
+                                          write_disposition="replace"))
+    return resource_list
+
+
+@dlt.resource(name="ticket_fields", write_disposition="replace")
+def ticket_fields_table(zendesk_client: Zenpy) -> Iterator[TDataItem]:
+    """
+    Dlt Resource that also writes to dlt state to update on custom ticket fields
+    @:param zendesk_client: Zenpy Object that is used to make API calls to Zendesk API
+    @:returns: Generator of dicts
+    """
+    # get dlt state
+    custom_tickets = dlt.state().setdefault("ticket_custom_fields", {})
+    # get all custom fields and update state if needed, otherwise just load dicts into tables
+    all_fields = zendesk_client.ticket_fields()
+    for field in all_fields:
+        field_id = str(field.id)
+        # grab id and update state dict if the id is new
+        if not (field_id in custom_tickets):
+            custom_tickets[field_id] = field.title
```

Comment:
here we we update state only with the initial value. this is fine (preserve schema). but IMO we should yield this initial value in `return_dict` so people have full information in the database

Reply:
As in add another key to the dictionary: "initial_field_name" : field_name?